[FLINK-7452] [types] Add helper methods for all built-in Flink types

twalthr opened the pull request:
## What is the purpose of the change

This PR provides an unified, detailed, and easy to use utility to work with Flink's built-in Java types.

This class helps users to get an overview about built-in Flink types and their features/limitations (esp. regarding null support).

## Brief change log

Improved `org.apache.flink.api.common.typeinfo.Types`.


## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

twalthr commented:
CC @alpinegizmo

twalthr commented:
@fhueske Feel free to review.

fhueske commented on flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java:
```diff
@@ -19,56 +19,408 @@
+	}
+
+	/**
+	 * Returns type information for Java arrays of primitive type (such as <code>byte[]</code>). The array
```
`The array and its elements do not support null values.` -> `The array must not be null.`? Entries of primitive arrays cannot be null anyway.

twalthr replied:
I just wanted to be consistent :D

fhueske commented on flink-examples/flink-examples-batch/src/main/scala/org/apache/flink/examples/scala/wordcount/WordCount.scala:
```diff
@@ -49,6 +51,8 @@ object WordCount {
 
     val params: ParameterTool = ParameterTool.fromArgs(args)
 
+    val t = Types.of[Either[String, Nothing]]
```
What's the purpose of this change?

twalthr replied:
Good question. Will remove it.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala:
```diff
@@ -20,60 +20,123 @@ package org.apache.flink.table.api
-  * This class enumerates all supported types of the Table API.
+  * This class enumerates all supported types of the Table API & SQL.
   */
 object Types {
```
Not sure if that's possible, but could the `Types` object extend the `JTypes` class such that we don't have to implement all methods again and forward the calls?

twalthr replied:
We had this initially. The problem is that Scala creates method where as Java fields. It would be a mixture of `STRING()`, `INTERVAL_MILLIS`. Furthermore, not all types of the Java API are important for Table API users. I think it is better to have specialized classes for each API. This also allows to have custom Javadocs. Like the STRING <-> VARCHAR mapping comment.

twalthr replied:
If you compare the two classes, you will notice that they are not the same. There are differences.

fhueske replied:
Makes sense, thanks

fhueske commented on flink-scala/src/main/scala/org/apache/flink/api/scala/Types.scala:
```diff
@@ -0,0 +1,363 @@
+  * use [[Types.of]] to let type information be determined automatically.
+  */
+@PublicEvolving
+object Types {
```
extend the Java `Types` class?

fhueske commented on flink-scala/src/main/scala/org/apache/flink/api/scala/Types.scala:
```diff
@@ -0,0 +1,363 @@
+  * types automatically. This class is intended for cases where the extraction is not possible
+  * (or inefficient) as well as cases where type information has to be supplied manually.
+  *
+  * Depending on the API you are using (e.g. Java API or Table API), there might be a more
```
Remove this comment?

twalthr commented on flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java:
```diff
@@ -19,56 +19,408 @@
+	 * (and non-final) or have a public getter and a setter method that follows the Java beans naming
+	 * conventions for getters and setters.
+	 *
+	 * <p>A POJO is a fixed-length, null-aware composite type with non-deterministic field order. Every field
```
Table API or a `keyBy(0, 1)` operation.

fhueske replied:
Key selection by index is not supported for POJOs and yields an exception. 
The case of importing a POJO DataStream as Table is described in the Table API / SQL documentation.

I would not add this detail here. IMO, it rather confuses users than helps them.

twalthr commented:
Thanks for the review @fhueske. I hope I addressed all your comments. I would not let the other classes extend the Java class. There are multiple differences. E.g. `Either` and `Option` in Scala/Java.

fhueske commented on flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java:
```diff
@@ -19,56 +19,417 @@
+ *
+ * <p>In many cases, Flink tries to analyze generic signatures of functions to determine return
+ * types automatically. This class is intended for cases where type information has to be
+ * supplied manually or would result in an inefficient type.
```
"or would result in an inefficient type" -> "or cases where automatic type inference results in an inefficient type."

fhueske commented on flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java:
```diff
@@ -19,56 +19,417 @@
+	 * A row itself must not be null.
+	 *
+	 * <p>A row is a fixed-length, null-aware composite type for storing multiple values in a
+	 * deterministic field order. Every field can be null independent of the field's type.
```
"independent of" -> "regardless of"?

twalthr commented:
@fhueske I updated this PR and added generic parameters as well.

twalthr commented:
@fhueske I would merge this later today if there are no objections.
